tests/unit/operator_logical_not.cpp

```diff
@@ -0,0 +1,83 @@
+
+/*
+ *   Copyright 2021 Huawei Technologies Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <sstream>
+
+#include <graphblas.hpp>
+
+using namespace grb;
+
+void grb_program( const size_t & n, grb::RC & rc ) {
+	(void)n;
+
+	{ // Test: logical_and< bool > just to make sure it works
+		bool value = true;
+		rc = rc ? rc : foldl( value, true, operators::logical_and< bool >() );
+
+		// Check results
+		if( value != true || rc != SUCCESS ) {
+			rc = FAILED;
+			std::cerr << "Test logical_and< bool > FAILED\n";
+			return;
+		}
+	}
+
+	{ // Test: logical_not< logical_and< bool > >
+		bool value = true;
+		rc = rc ? rc : foldl( value, true, operators::logical_not< operators::logical_and< bool > >() );
+
+		// Check results
+		if( value != false || rc != SUCCESS ) {
+			rc = FAILED;
+			std::cerr << "Test logical_not< logical_and< bool > > FAILED\n";
+			return;
+		}
+	}
```

Comment:
Three comments:
1. should we not also test negation of or and xor (do we even have xor actually?)
2. should we not test for false associative property of nand and nor?
3. should we not test the full truth table (two Booleans that can each take four values, i.e., four tests per operator)?

For bonus points, I don't think other operators have this exhaustively tested though, maybe all of foldl, foldr, and apply should be tested. Maybe with some nice software engineering this is actually doable with not too much repeated code

Reply:
> Three comments:
> 
>     1. should we not also test negation of or and xor (do we even have xor actually?)

We don't have xor, but why not.

>     2. should we not test for false associative property of nand and nor?

Absolutely.

>     3. should we not test the full truth table (two Booleans that can each take four values, i.e., four tests per operator)?

Will do.

> For bonus points, I don't think other operators have this exhaustively tested though, maybe all of foldl, foldr, and apply should be tested. Maybe with some nice software engineering this is actually doable with not too much repeated code

I'll see what I can think of but it would be nice to have such testing.